### ISSUE TITLE
[sklearn#33646] [Callbacks] Should/Can we avoid creating the context tree when no callbacks ?

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#384]
**Mirrored from [scikit-learn#33646](https://github.com/scikit-learn/scikit-learn/issues/33646)**

Original title: [Callbacks] Should/Can we avoid creating the context tree when no callbacks ?
Original labels: Callbacks
Created: 2026-03-27T00:03:17Z

---

For meta-estimators, the answer is a definitive no. They can have subestimators with callbacks which need the full context tree to work optimally.

For regular estimators, it could make sense. The performance for creating contexts is not prohibitive: ~1-2us. This is small for one iteration even for quick estimators and small datasets (knowing that callbacks don't go deeper than iterations involving the full dataset). But there might be cases where it matters.

I would avoid doing it if it adds too much verbosity within every estimator, like adding conditional branches everywhere we create a subcontext or call callback hooks. But if we can enable the feature without changing the existing callback support api implementation, why not.

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to [11](https://github.com/godofecht/flow-scikit/issues/11) in the flow-scikit repository. Further library-specific triage and implementation will happen there.
